### PR TITLE
plugin/sql backend provider

### PR DESCRIPTION
**Pull request by thmeitz**
### 1. Why is this pull request needed and what does it do?

- supports many relational databases (see plugin README.md)
- supports many resource record types (35 RRs are supported - please check which ones does'nt makes sense)

Please be aware, that most of the code in the ent directory was auto generated by [entgo](https://entgo.io/docs/getting-started). The needed rr record methods are also autogenerated.

It does not get less code! That's the price of having everything in a reasonable ORM, including validation, hooks, etc.

Furthermore I have stripped the example code for the zone import. Look into the utils directory how records and zones can be imported into the database. 

All Ariga Atlas related things are removed. So you have to handle the migrations by yourself (or use the automigrate feature in the Corefile). The SQL files containing the ddl statements are located in the migration directory for MySQL, MariaDB and PostgreSQL.

Outlook: I plan to provide a Relay compatible GraphQL Interface based on the current plugin that can be used to manage both zones and RRs. Without a management API the sql plugin is pretty useless and you have to import the zones by yourself.

### 2. Which issues (if any) are related?

#1663 

### 3. Which documentation changes (if any) need to be made?

README.md needs a change, if you accept that sql is living in the plugin directory

### 4. Does this introduce a backward incompatible change or deprecation?
No.

### CODE REVIEW

File: plugin/sql/utils/importrecord.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+type DnsRecordResource interface {
+	Header() *dns.RR_Header
+}
+
+// ImportRecord imports a zone, the RR_header and the zone_record into the database
+// the zone_record has to be a json string marshalled with New<RR>().Marshal().
+// The zone must exists.
+func ImportRecord[K DnsRecordResource](client *ent.Client, zone string, i K, zone_record string) error {
```
[the remark this answers is not in the document]
Comment (thmeitz) [reply plugin/sql/utils/importrecord.go:18]: Stickler does not like generic methods in Go?

**Review by dilyevsky — COMMENTED** [review]
IMO given absolutely massive size and esoteric tooling this should be a standalone (external) plugin.

**Comment by thmeitz** [comment]
It is a pity that this is how it is seen. If a plugin is only allowed to have 1000 lines so be it.

Btw, the esoteric software holds more Github stars than CoreDNS. Deal with the software before you judge.

Sorry for disturbing you guys.

I close the PR.

**Comment by LtdJorge** [comment]
@thmeitz do you happen to have this as an external plugin?

**Comment by thmeitz** [comment]
@LtdJorge no, not as an external plugin.

**Pull request closed without merging**